The 0-ticking Patch (1.1)

Lexer747 opened the pull request:
# The 0-ticking Patch (1.1)

* New support for getting the correct weapon stats when switching
  weapons on the same tick that a player attacks.
* Adds Salamander Support 🎉.
* Adds Harmonized 4t Support (only triggers when using standards).
  * Further improvements to powered staff detection in general required to enable harm.
* Basic Red Keris special attack support.


I do understand this is a big patch, therefore from the get go, please just say if you want this split into smaller more reviewable PRs. Essentially it could be reasonably broken down into one PR per top-level bullet point above, I am more than happy to do this work if you would like. Please just say!

## Implementation details:

The quick swapping support (known as 0-ticking in the PvP community), is a helpful feature I wanted for rooms like Nylos and Baba-puzzle where many frequent and fast swaps happen. I noticed that the underlying cause of the problem was that we are essentially caching the results of `getWeaponStats()` and related functions when the `onGameTick()` subscription was triggered. Unfortunately this subscription happens at the start of a tick so you could very easily create a bug like so: (Youtube link: https://youtu.be/t_2h5uJM5kk or uploaded to [github](https://github.com/ngraves95/attacktimer/assets/17512637/b075d68b-c508-401f-9174-27a79b6fffc6) as an `.mp4`). As you can see if you attack and swap on the same tick the wrong cool down is triggered, e.g. blowpiping with a 5t timer above my head.

This change tackles the root of the problem by using a new a subscriber to listen a client int `384` which from testing will always monotonically go up as the client switches any worn item. This means we can use a fresh non-stale weapon id for the equipped weapons stats. This required some more finicky details, including a new state for the first tick of attack delay, this is so the plugin can only do a "weapon" swap. If it's early enough to not have already been rendered.

------------------------------------------------------------------------

Next up is Salamander support, which turned out to just need a spot anim. As it turns out the ranged and magic attacks have no animation at all and simply trip up the `isPlayerAttacking()` code to think no attack was happening.

------------------------------------------------------------------------

Finally the even more improved staff support, as I was fixing the 0-ticking I was noticing quite a few problems with d97172936b4e8f82ecdbb1666e9cd2634aa50e30 when staffs were involved. Especially when quickly swapping and manual casting. So I tackled it from a bunch of angles trying to reduce the false positives and false negatives that come from the current solution.

This starts with the new `PoweredStaves.java` file (which renamed from `WeaponIds.java` but git has lost the rename), the first step of this is too explicitly link every staff to it's corresponding animation and projectile. With this new information we can then already be more accurate by only switching to 4t when the animation and weapon Id match, whereas the old version just looked at id and animation separately.

Next there's some new data `Spellbook.java` and `CastingSoundData.java` which allows the plugin to correlate which spell book the player is on and if that animation can be manually cast. Giving a us a really big win in 99% of cases. This is because most things people do they're on thralls/ancients/lunars and all the powered staves tend to use the standard spellbook animations. We also use the sound to help detect manual casts, but a lot is still left on the table here as we should also be detecting the sound of the staff being used too, but I haven't got around to implementing that.

The other reason I haven't worked on the sound side of it too much is that it doesn't work when the client is muted so it feels half-baked in premise.

The last piece of the staff puzzle is the most controversial because it uses the deprecated API `getProjectiles()`, now it's not documented why this API is deprecated or the replacement to use instead, so my gut says this patch may need this part being ripped out because this is breaking some jagex rules. And for the record I'm more than ok ripping it out, due to how to ends up working. Which I'll explain now:

This last piece is that using the projectiles we can check on the next tick what projectile the player fired, if it doesn't match the staff then they must have been casting a spell instead. Note that this whole section is what enabled Harm as we can tie to specific animations and projectiles to be sure the player is casting something which will be upgraded to only 4 ticks instead of 5, like curse or bind.

------------------------------------------------------------------------

### Misc details:
 * Version bump the whole plugin to 1.1 since this a big change. New subscribers, different state management, etc.
 * Fixed a small bug with blowpipe tailing animation being picked up as a second attack (see inline comment in `isPlayerAttacking()`.
 * Added @ngraves95 to more copyright headers. (I've just assumed it's all copyright circa 2022 because checking the history was a touch too tedious for me).
 * Added some very basic sanity check unit tests which while adding I discovered that I couldn't run them locally. This is because Runelite has moved to `lombok:1.18.30` instead of `lombok:1.18.20`, this is a breaking change with the V7.4 of gradle (Fails with:    `java.lang.IllegalArgumentException: Unsupported class file major version 65`).  Therefore to match the lombok version we just have to bump the gradle version to 8.X.

# Testing

I test a bunch of random weapons here https://www.youtube.com/watch?v=TC6O1SSEwTw if you would like me to test any more senarios or ideas you have just let me know! More than happy to verify things are good.

Lexer747 commented:
@ngraves95 please could you have a look at your convenience, thanks in advance :)

Lexer747 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -220,32 +415,75 @@ private boolean isPlayerAttacking()
+        return fromId != null;
+    }
+
+    private boolean getIsNotFishingSpot(Actor target) {
+        final NPC npc = (NPC) target;
+        FishingSpot spot = FishingSpot.findSpot(npc.getId());
+        return spot == null;
     }
```
Most recent commit f1869aa | Saw someone raise this on the runelite discord:

> https://imgur.com/rcZoS5G
> any idea why this has started happening with attack timer metronome? haven't changed anything afaik

This small function fixes it:


https://github.com/ngraves95/attacktimer/assets/17512637/aa166886-11e8-4650-9253-16cc58f1cbc9

ngraves95 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -416,7 +417,9 @@ private boolean isPlayerAttacking()
-            boolean isTargetDummy = target.getHealthRatio() == -1 && target.getCombatLevel() == 0;
+            // Turns out fishing spots look like target dummies... so check it's not one
+            boolean isNotFishingSpot = getIsNotFishingSpot(target);
+            boolean isTargetDummy = target.getHealthRatio() == -1 && target.getCombatLevel() == 0 && isNotFishingSpot;
```
Is there a list of things that have a health ratio? I think this is a potential source of false postivies.

 I've noticed false positives on the attack countdown when interacting with non-attackable NPCs, usually while on a casting cooldown. For example, I was on high alch cooldown and the attack timer triggered when I got a new Mahogany Homes contract from Amy.

ngraves95 replied:
Granted, false positives are preferable to false negatives if we have to make a tradeoff.

Lexer747 replied:
> For example, I was on high alch cooldown and the attack timer triggered when I got a new Mahogany Homes contract from Amy.

Oh wow thats quite unfortunate. Yeah I can easily reproduce that by just walking towards an NPC with any non-zero animation. We could take inspiration from the xp tracker plugin which needs to find exp modifiers:

```java
if (interacting instanceof NPC && COMBAT.contains(skill))
{
	final int xpModifier = worldSetToType(client.getWorldType()).modifier(client);
	final NPC npc = (NPC) interacting;
	xpState.updateNpcExperience(skill, npc, npcManager.getHealth(npc.getId()), xpModifier);
}
```

Here the `getHealth()` might be more reliable that health ratio?

There's also an option to explore which is looking for an attack option in the menu options of a given NPC. I'll have a play with both...

Lexer747 replied:
So using the menu and hard coding the dummy_id works much better defensively reducing the false positives while not affecting the false negatives. I'm going to test it a bit more before committing it and pushing. 

This can still be triggered by alching and interacting with an NPC which is attackable, but this would fix the fishing spots, Amy and that class of interacting issues (some pick-pocketing, etc) with NPCs which don't have health and levels. 

It still leaves me unsatisfied because pick-pocketing knights, etc will still be an "attack". We could go with a block list here? A list of animations we know which aren't attacks for the rarer NPCs which have attack options but can also be interacted with which isn't an attack.


``` diff
diff --git a/src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java b/src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java
index 853046a..0855c03 100644
--- a/src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java
+++ b/src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java
@@ -32,2 +32,4 @@ import java.awt.Color;
 import java.awt.Dimension;
+import java.util.Arrays;
+
 import javax.inject.Inject;
@@ -59,2 +61,3 @@ import net.runelite.client.game.FishingSpot;
 import net.runelite.client.game.ItemManager;
+import net.runelite.client.game.NPCManager;
 import net.runelite.client.plugins.Plugin;
@@ -102,2 +105,5 @@ public class AttackTimerMetronomePlugin extends Plugin
 
+    @Inject
+	private NPCManager npcManager;
+
     public int tickPeriod = 0;
@@ -410,2 +416,4 @@ public class AttackTimerMetronomePlugin extends Plugin
 
+    private static final int TARGET_DUMMY_ID = 10507;
+
     private boolean isPlayerAttacking()
@@ -419,6 +427,7 @@ public class AttackTimerMetronomePlugin extends Plugin
         {
-            // Turns out fishing spots look like target dummies... so check it's not one
-            boolean isNotFishingSpot = getIsNotFishingSpot(target);
-            boolean isTargetDummy = target.getHealthRatio() == -1 && target.getCombatLevel() == 0 && isNotFishingSpot;
-            boolean attackingNPC = target.getCombatLevel() > 0 || isTargetDummy;
+            final NPC npc = (NPC) target;
+            boolean containsAttackOption = Arrays.stream(npc.getComposition().getActions()).anyMatch("Attack"::equals);
+            Integer health = npcManager.getHealth(npc.getId());
+            boolean hasHealthAndLevel = health != null && health > 0 && target.getCombatLevel() > 0;
+            boolean attackingNPC = hasHealthAndLevel || npc.getId() == TARGET_DUMMY_ID || containsAttackOption;
             // Not walking is either any player animation or the edge cases which don't trigger an animation, e.g Salamander.
@@ -441,6 +450,5 @@ public class AttackTimerMetronomePlugin extends Plugin
 
-    private boolean getIsNotFishingSpot(Actor target) {
-        final NPC npc = (NPC) target;
+    private boolean getFishingSpot(NPC npc) {
         FishingSpot spot = FishingSpot.findSpot(npc.getId());
-        return spot == null;
+        return spot != null;
     }


```

Lexer747 replied:
2t skilling methods are also affected by this issue but atleast that one sort of makes sense 😅

Lexer747 replied:
[03f6a3a](https://github.com/ngraves95/attacktimer/pull/56/commits/03f6a3a3e7701e21be43117c263f27f3d85c9887) is my best attempt at reducing false positives, it introduces a block list of animations (only pick-pocket on that list for now). 

Happy to revert the block list if you think that direction is poor. There's still ways to trigger the timer incorrectly, like being hit (flinching animation) then interacting, or alching while interacting with an NPC with a health bar (talking, still walking to attack the NPC). Having no ammo in the quiver.

Lexer747 replied:
Testing it also fixes https://github.com/ngraves95/attacktimer/issues/55 though off the back of this testing I've added flinching to the exclude list because that frequently mis-triggers it. 4c0c29f

Just say if you want me to rebase/squash to one commit.

ngraves95 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -25,8 +71,8 @@ public class AttackTimerMetronomePlugin extends Plugin
 {
     public enum AttackState {
         NOT_ATTACKING,
+        DELAYED_FIRST_TICK,
```
Is this state when you are flinched?

Lexer747 replied:
This is not the state when you are flinched as you figured out. 

This is the state we set for the first tick of any attack. This is the only state in which the attack speed can be update more than once, through the `onVarClientIntChanged` subscriber. As if the state is not `DELAYED_FIRST_TICK` then `onVarClientIntChanged` cannot change the speed of the weapon. 

But if the state is `DELAYED_FIRST_TICK`  then the subscriber will re-call `performAttack()` hence updating the speed of the weapon as the player swaps.

ngraves95 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -60,15 +109,73 @@ public enum AttackState {
     public int attackDelayHoldoffTicks = ATTACK_DELAY_NONE;
 
     public AttackState attackState = AttackState.NOT_ATTACKING;
+    // The state of the renderer, will lag a few cycles behind the plugin's state
```
Why is that?

Lexer747 replied:
Sorry I should have also been clearer here.

"cycles" in this comment refers to the `client.getGameCycle()` method

``` java
	/**
	 * Gets the local clients game cycle.
	 * <p>
	 * Note: This value is incremented every 20ms.
	 *
	 * @return the game cycle
	 */
	int getGameCycle();
```

So the comment "The state of the renderer, will lag a few cycles behind the plugin's state" is not talking about ticks. The reason for this lag in the rendered compared to the state is that `onGameTick` will occur very early in overall Tick, about 1-5 cycles (out of the 30 cycles in a single tick), so the state is set early here. Then very late at the end of that game tick the render override is called `@Override public Dimension render(Graphics2D graphics)()` - something like the 15-25th cycle in a given tick. Take these numbers with a grain of salt, and I can't explain the ordering because it's somewhere in the runelite guts.

This de-sync is a great heuristic for when it's too late in a given tick for a weapon swap to take effect. As by the time the render is called any weapon swaps will happen on the next tick. But any swaps noticed by `onVarClientIntChanged` which happen before the renderer is insync will be happening right now and affect this tick. Hopefully that explanation helps! Let me know if you want it codified as a comment into the plugin.

ngraves95 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -174,42 +281,129 @@ private int adjustSpeedForLeaguesIfApplicable(int baseSpeed)
+        // can do slightly better by checking the projectile we fired, as the powered staves have custom
+        // projectiles. But we start in the above getWeaponSpeed call to just assume holding the staff means
+        // 4-tick.
+        Integer firedProjectile = getPlayerFiredProjectile();
```
I would prefer to not depended on a deprecated API. Can we remove the projectile tracking code?

Lexer747 replied:
Affirmative, I will push a commit removing it 🫡

ngraves95 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -220,32 +414,67 @@ private boolean isPlayerAttacking()
             boolean attackingNPC = target.getCombatLevel() > 0 || isTargetDummy;
-            boolean notWalking = animationId != -1;
+            // Not walking is either any player animation or the edge cases which don't trigger an animation, e.g Salamander.
+            boolean notWalking = animationId != -1 || getSalamanderAttack();
```
I think this is check for animationId != -1 is causing false positives with the player being flinched. Can we add an exclude list of common animations that are not attacks?

Lexer747 replied:
Yeah it definitely isn't perfect, but I can say it's already paying off. All the new Emberlight, Bone claws, etc just work. No new code needed for them to be picked up just fine.

Adding more exclusions will only help, flinching/eating and thralls will go along way to ensuring less false positives occur.

Lexer747 replied:
![java_9ungKII5gx](https://github.com/user-attachments/assets/c64af185-8cfb-4c6a-a184-d14ad27db50c)
![java_MGCzYMoOe9](https://github.com/user-attachments/assets/666212c1-176e-41f5-a941-422a3e538d9a)

ngraves95 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -408,19 +412,26 @@ private int getWeaponSpeed(int weaponId, Integer firedProjectile, PoweredStaves
         // proof to new weapons which don't need custom code and the weapon
         // stats are enough.
         Actor target = client.getLocalPlayer().getInteracting();
         if (target != null && (target instanceof NPC))
```
Does this break the timer for PvP?

Lexer747 replied:
Doh, NGL I didn't test PvP it's something I rarely do. I'll test it today and make fixes if it's required.

ngraves95 commented:
Nice improvements - please also squash before committing. I think a blocklist of common non-attack animations is the right call here. mostly looks good but a few things:
* Let's make sure that checking instanceof NPC doesn't break PvP/LMS.
* I'd prefer not to use the projectile tracking sections if possible because the API is deprecated.

Lexer747 commented:
Good spot on the PvP thing (and thanks for the review in general), it not only doesn't work, but needs a separate branch because we can't check a player object for all the things we use on NPCs (Attack option, etc). However the fix is relatively easy, see the latest version of the code.

Before | After
---|---
![ApplicationFrameHost_FEn7wBKHV8](https://github.com/user-attachments/assets/daef8f3b-72c3-47fe-ad71-fb4d65753013)|![java_SMo3Ga4YbC](https://github.com/user-attachments/assets/547f9d68-c02e-4e38-8151-f583c1431140)

The projectile code is now also removed (just the deprecated code and calling functions), I've left the data in just in case we ever add it back in.

This PR now squashed into one commit.

ngraves95 commented on src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java:
```diff
@@ -60,15 +108,75 @@ public enum AttackState {
 
     public AttackState attackState = AttackState.NOT_ATTACKING;
+    // The state of the renderer, will lag a few cycles behind the plugin's state. "cycles" in this comment
+    // refers to the client.getGameCycle() method, a cycle occurs every 20ms, meaning 30 of them occur per
```
Kudos: very helpful comment about the difference between cycles and ticks here.

ngraves95 commented:
I was just testing Trident of the Swamp and its showing up as 5t. Let's double check there are no regressions on powered staves after removing the projectile tracking code

Lexer747 commented:
> I was just testing Trident of the Swamp and its showing up as 5t. Let's double check there are no regressions on powered staves after removing the projectile tracking code.
> 
> I think it might be because the tridents use a standard spellboook wave animation.

Which spell book where you on when it was 5t? (standards?)

I think I've realised the regression regardless which is that if the spell book does match the stave built-in spell it now defaults to a manual cast instead of the inbuilt spell 🤦‍♂️ That should be reversed to match the old method. We could do better here by matching the sound of the staff built in spell but that still isn't 100% reliable because it doesn't work when sound is off.

The fix is probably just this, because this check only made sense when we had the projectile code in place. Let me test it though.

``` diff
@@ -309,7 +309,7 @@ public class AttackTimerMetronomePlugin extends Plugin

     private int getWeaponSpeed(int weaponId, PoweredStaves stave, AnimationData curAnimation, boolean matchesSpellbook)
     {
-        if (stave != null && !matchesSpellbook)
+        if (stave != null)
         {
             // We are currently dealing with a staves in which case we can make decisions based on the
             // spellbook flag. We can only improve this by using a deprecated API to check the projectile
```

ngraves95 commented:
> I was just testing Trident of the Swamp and its showing up as 5t. Let's double check there are no regressions on powered staves after removing the projectile tracking code.
> 
> I think it might be because the tridents use a standard spellboook wave animation.

Ok, I managed to fix this by adding a check for `stave.getAnimations().contains(curAnimation)`. Also validated that manual casting still shows up as 5 tick (earth bolt manual cast-> 5t, trident of the swamp autocast -> 4t).

I'll merge this and then add the fix locally.

I'm still seeing some issues where pickpocketing while in an alch animation triggers the timer but that feels like a low priority. I'll get around to adding alch and other utility spells to the blocklist

Lexer747 commented:
Oh yeah that's actually a better fix than my small diff, slightly more strict on which animations can trigger this false positive - I like it. 

Ah - Alching while pick-pocketing! I didn't think of that 😆

Lexer747 commented:
092130b (the `stave.getAnimations().contains(curAnimation)` fix) testing:

Builtin ✅ | Wind Blast ✅ | Fire Wave ❌
---|---|---
![java_aic9V4ZHfj](https://github.com/user-attachments/assets/60f1d182-3b94-41a5-ac63-7b677d5119ca)|![java_ukS7m8dc0c](https://github.com/user-attachments/assets/43f732c7-218c-4409-882c-6699ff2d00ae)|![java_sirk3EfbQL](https://github.com/user-attachments/assets/85b57237-1ac2-4e75-84c6-70245da1c251)

ngraves95 commented:
> [092130b](https://github.com/ngraves95/attacktimer/commit/092130b81401a4e12acd8f33139c6fea02326889) (the `stave.getAnimations().contains(curAnimation)` fix) testing:
> 
> Builtin ✅	Wind Blast ✅	Fire Wave ❌
> ![java_aic9V4ZHfj](https://github.com/user-attachments/assets/60f1d182-3b94-41a5-ac63-7b677d5119ca)	![java_ukS7m8dc0c](https://github.com/user-attachments/assets/43f732c7-218c-4409-882c-6699ff2d00ae)	![java_sirk3EfbQL](https://github.com/user-attachments/assets/85b57237-1ac2-4e75-84c6-70245da1c251)

I dont think there's anything we can do about wave animations colliding with powered staves right now until we get an non-deprecated projectile tracking API. The overlap of people casting wave spells while using a powered staff should hopefully be almost 0, though.
